clustermesh: cli: output cluster values to dict conditionally

MrFreezeex opened the pull request:
Fixes https://github.com/cilium/cilium/issues/40796
Followup to https://github.com/cilium/cilium/pull/40879

First commit allow disabling ReuseValues on helm upgrade and second commit to output `clustermesh.config.clusters`  to a dict if running Cilium >= 1.20. See the commit description for a bit more details.

```release-note
clustermesh: cli: support outputting `clustermesh.config.clusters` to the new dict format if the cluster is running Cilium >= 1.20
```

Test PR to output in the dict format in >= 1.19: https://github.com/cilium/cilium/pull/40960

MrFreezeex committed:
clustermesh: cli: use ReuseValues false for helmUpgrade

This commit adds support for setting ReuseValues to false in the
helmUpgrade function. This is helpful to prepare for outputting
`clustermesh.config.clusters` to a dict in the CLI and prevent
uncontrolled deep merged if we were using ReuseValues.

Since we now only need the clusters map and not the "full" values this
commit rename a lot of code to replace most usage of "values" by
"clusters".

Signed-off-by: Arthur Outhenin-Chalandre <git@mrfreezeex.fr>

MrFreezeex committed:
clustermesh: cli: output cluster values to dict conditionally

Output `clustermesh.config.clusters` to a dict if the cluster is running
at least Cilium 1.20. We are still output-ing in a list form since the
dict form will be supported from Cilium 1.19 and we keep one additional
minor version with the list form to allow users to downgrade and reusing
the same values.

Signed-off-by: Arthur Outhenin-Chalandre <git@mrfreezeex.fr>

marseel commented on cilium-cli/clustermesh/clustermesh.go:
```diff
@@ -1176,31 +1176,31 @@ func (k *K8sClusterMesh) validateCAMatch(aiLocal, aiRemote *accessInformation) (
-	remoteNewCluster       map[string]any                 // local cluster to add to remote clusters
-	remoteClusterNames     []string                       // names of remote clusters for displaying logs
-	remoteClusterNamesAi   []string                       // names of remote clusters for remove sections
+	localOldClusters     map[string]any                 // current clusters values
```
I know it's not your fault, but this structure really needs refactor, it's so hard to understand what is what :sweat_smile: 
I'm thinking about something like:
```
type ClusterMeshState struct {
  localClusterState ClusterState
  remoteClusterStates map[string]ClusterState
}
```
But it's not for this PR, just mentioning it in case you would have free cycles.

MrFreezeex replied:
Yep I am not exactly sure if this would fully solve all the case but it would be nice indeed to find ways to do something along those lines + perhaps rely more on function args that this "global" state

marseel commented on cilium-cli/clustermesh/clustermesh.go:
```diff
@@ -1176,31 +1176,31 @@ func (k *K8sClusterMesh) validateCAMatch(aiLocal, aiRemote *accessInformation) (
-	remoteClusterNamesAi   []string                       // names of remote clusters for remove sections
+	localOldClusters     map[string]any                 // current clusters values
+	localNewClusters     map[string]any                 // new clusters values
+	localClusters        map[string]any                 // localOldClusters + localNewClusters => local clusters values
```
This `localClusters` can be removed, no? It's assigned within `connectLocalWithHelm` and then never used again?

MrFreezeex replied:
Yep look like it, good catch, thanks!

marseel commented on cilium-cli/clustermesh/clustermesh.go:
```diff
@@ -1279,73 +1280,96 @@ func (k *K8sClusterMesh) processRemoteClients(ctx context.Context, remoteClients
+	if err != nil {
+		return fmt.Errorf("Failed to parse Helm chart version %s on cluster %s: %w", release.Chart.Metadata.Version, client.ClusterName(), err)
+	}
+	if versioncheck.MustCompile("<1.20.0")(version) {
```
shouldn't that be `<1.19.0`? We want to use new version for 1.19.

MrFreezeex replied:
The <1.20.0 is essentially to get one additional version that users can downgrade with the same values as they were previously using in the previous version ~. @giorio94 suggested me to do this here https://github.com/cilium/cilium/issues/40796#issuecomment-3135481234

marseel replied:
Makes sense, thanks!

marseel commented on cilium-cli/clustermesh/clustermesh.go:
```diff
@@ -1279,73 +1280,96 @@ func (k *K8sClusterMesh) processRemoteClients(ctx context.Context, remoteClients
+	if versioncheck.MustCompile("<1.20.0")(version) {
+		clustersRaw = convertClustersToListClusterMesh(clusters)
+	}
+	err = setClustersInValues(release, client, clustersRaw)
```
I think we have `getEnabledClusters` that would be removing all clusters that have "enabled: false". I think a bit more correct way would be to not filter them anymore with `getEnabledClusters` and only here filter depending on version. That would ensure in case of connecting clusters with new version, we won't strip disabled clusters. It would be great to add unit test covering it as well.

MrFreezeex replied:
I thought about doing something along these lines too but this would also necessitate that we add some enabled=false check in many places in this code base and this sounds a bit scary + would complexify a bit further the code  :sweat:. My reasoning here is that disabled cluster was mostly a helm feature (useful when overriding thing between values) and that it was ok to strip when using the clustermesh CLI BUT it's true that theoretically there could be some users that doesn't install Cilium as code and disable some clusters using that... But if we were to do that we should also probably keep that for the list format too (as it support disabled clusters too ~).

One thing that we could theoretically do though is to save disabled clusters in the state aside and reintroduce it somewhere here essentially which would be easier as all the functions wouldn't have to support disabled clusters (I am not sure how I would be able to unit test that yet but maybe I will find something somehow).

MrFreezeex replied:
So I did the above in some separate commit, let me know if the approach looks right to you!

marseel reviewed (changes requested):
Thanks and apologies for late review!

MrFreezeex committed:
clustermesh: cli: remove localClusters field from global state

Signed-off-by: Arthur Outhenin-Chalandre <git@mrfreezeex.fr>

MrFreezeex committed:
clustermesh: cli: reintroduce disabled clusters on helm upgrade

Signed-off-by: Arthur Outhenin-Chalandre <git@mrfreezeex.fr>

marseel commented on cilium-cli/clustermesh/clustermesh.go:
```diff
@@ -1486,7 +1496,7 @@ func (k *K8sClusterMesh) connectRemoteWithHelm(ctx context.Context, localCluster
 			defer func() { <-sem }()
 
-			if err := k.connectSingleRemoteWithHelm(ctx, rc, cn, helmVals); err != nil {
+			if err := k.connectSingleRemoteWithHelm(ctx, rc, cn, helmVals, state.localDisabledClusters); err != nil {
```
This is incorrect I think? We should be passing remote Disabled clusters here. no?

MrFreezeex replied:
yep indeed :man_facepalming: I fixed all occurrence of this and found one thing in the state that can be removed (`remoteClustersDelete` as it's used in two functions right to each other...)

marseel commented on cilium-cli/clustermesh/clustermesh.go:
```diff
@@ -1176,35 +1177,34 @@ func (k *K8sClusterMesh) validateCAMatch(aiLocal, aiRemote *accessInformation) (
-		remoteOldClustersAll:   make(map[string]map[string]any),
-		remoteHelmValuesDelete: make(map[*k8s.Client]map[string]any),
-		remoteNewCluster:       make(map[string]any),
+		localOldClusters:      make(map[string]any),
```
`remoteOldDisabledClusters` is not initialized. CI failures seem to be related to it: https://github.com/cilium/cilium/actions/runs/17084338553/job/48445792499

MrFreezeex replied:
Woops sorry I fixed that :man_facepalming:! Thanks!

Because it's the last PR to finish this overall feature I just rebased and fixed something else related to the helm code that wasn't properly excluding disabled clusters in a separate commit. Note that from the existing commit you reviewed I only made this one change about initializing the map correctly.

MrFreezeex committed:
clustermesh: helm: fix disabled cluster exclusions

The default helm function was not working with disabled cluster, let's use a
not equal to explicitly remove disabled clusters (note that clusters are
enabled by default unless explicitly disabled).

Also adding a deepCopy to make sure we are not mutating the existing
values.

Signed-off-by: Arthur Outhenin-Chalandre <git@mrfreezeex.fr>

marseel reviewed (approved):
Thanks!
